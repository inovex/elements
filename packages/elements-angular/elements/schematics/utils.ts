import { virtualFs, workspaces } from '@angular-devkit/core';
import { SchematicsException, Tree } from '@angular-devkit/schematics';

const CONFIG_PATH = 'angular.json';

function createHost(tree: Tree): workspaces.WorkspaceHost {
  return {
    async readFile(path: string): Promise<string> {
      const data = tree.read(path);
      if (!data) {
        throw new SchematicsException('File not found.');
      }
      return virtualFs.fileBufferToString(data);
    },
    async writeFile(path: string, data: string): Promise<void> {
      return tree.overwrite(path, data);
    },
    async isDirectory(path: string): Promise<boolean> {
      return !tree.exists(path) && tree.getDir(path).subfiles.length > 0;
    },
    async isFile(path: string): Promise<boolean> {
      return tree.exists(path);
    },
  };
}

export function readConfig(host: Tree) {
  const sourceText = host.read(CONFIG_PATH)?.toString('utf-8') || '';
  return JSON.parse(sourceText);
}

export function writeConfig(host: Tree, config: JSON) {
  host.overwrite(CONFIG_PATH, JSON.stringify(config, null, 2));
}

export function getAngularAppConfig(
  config: any,
  projectName: string
): any | never {
  if (!config.projects.hasOwnProperty(projectName)) {
    throw new SchematicsException(`Could not find project: ${projectName}`);
  }

  return config.projects[projectName];
}

export async function getProjectName(tree: Tree) {
  const host = createHost(tree);
  const { workspace } = await workspaces.readWorkspace('/', host);
  return workspace.extensions.defaultProject as string;
}
