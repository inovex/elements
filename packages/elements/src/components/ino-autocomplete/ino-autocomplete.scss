@use '../base/mdc-customize';
@use '@material/list';
@use '../base/theme' as theme;
@use '../base/new-theme' as new-theme;

@include list.deprecated-core-styles;

.ino-autocomplete__list {
  margin: 0.5rem 0 0 0;
  padding: 0;
  background-color: white;
  box-shadow: 0px 0px 22px 0px rgba(93, 91, 105, 0.35);
  border-radius: 10px;
  overflow: auto;
}

.ino-autocomplete__list-item {
  height: 1.5rem;
  font-size: 0.85rem;
  padding: 7px 0 7px 9px;

  &--selected, &:hover {
    background-color: new-theme.$p-1;
    color: black;
  }

  &--highlight {
    color: theme.color(primary, dark);
    background-color: transparent;
    font-weight: bold;
  }

  &--no-match {
    color: theme.color(light, dark);
    pointer-events: none;
  }
}

.ino-input--font-grey {
  --ino-input-color: #{new-theme.$n-5};  //n-05
}