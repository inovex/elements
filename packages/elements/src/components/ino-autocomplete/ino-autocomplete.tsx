import {
  Component,
  ComponentInterface,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Listen,
  Prop,
  State,
  Watch,
} from '@stencil/core';
import classNames from 'classnames';
import {
  getSlotContent,
  hasSlotContent,
  moveCursorToEnd,
} from '../../util/component-utils';
import { Debouncer } from '../../util/debouncer';

const NO_OPTION_SELECTED = -1;

/**
 * @slot input - An `<ino-input>` element that will be controlled by this component
 * @slot default - A list of `<ino-option>` elements as options
 */
@Component({
  tag: 'ino-autocomplete',
  styleUrl: 'ino-autocomplete.scss',
  shadow: true,
})
export class Autocomplete implements ComponentInterface {
  private inputEl: HTMLInoInputElement;
  private optionEls: HTMLInoOptionElement[];
  private menuContainer: HTMLDivElement;
  private lastKeyCode: string;
  private debouncer: Debouncer = new Debouncer();
  private filteredOptionEls: HTMLInoOptionElement[];
  private selectedOption?: HTMLInoOptionElement;
  private _selectedOptionIndex?: number = NO_OPTION_SELECTED;

  private set selectedOptionIndex(index: number) {
    if (this.selectedOption) {
      this.selectedOption.selected = false;
    }

    this._selectedOptionIndex = index;
    this.selectedOption = index >= 0 ? this.optionEls[index] : undefined;

    if (this.selectedOption) {
      this.selectedOption.selected = true;
    }
  }
  private get selectedOptionIndex(): number {
    return this._selectedOptionIndex;
  }

  @Element() el: HTMLInoAutocompleteElement;

  @State() private menuIsVisible = false;
  @State() private noOptionsIsVisible = false;

  /**
   * Timeout of the debouncing mechanism used when filtering the options.
   */
  @Prop() debounceTimeout = 300;

  /**
   * Text to display when there are no options found.
   */
  @Prop() noOptionsText = 'No Option';

  /**
   * Value of the autocomplete
   */
  @Prop() value;

  @Watch('value')
  onValueChange(newValue, oldValue): void {
    if (newValue === oldValue) {
      return;
    }
    this.setOptionByValue(newValue);
  }

  /**
   * Emits in three ways:
   *
   * 1. Clicking on an option
   * 2. Pressing `Enter` while an option is selected
   * 3. Entering a valid value and blurring the input element
   *
   * Contains one of the texts provided by the `<ino-options>`s.
   */
  @Event() valueChange: EventEmitter<string>;

  componentDidLoad(): void {
    this.setupInput();
  }

  disconnectedCallback(): void {
    this.inputEl.removeEventListener('inoFocus', this.onInoInputFocus);
    this.inputEl.removeEventListener('inoBlur', this.onInoInputBlur);
  }

  @Listen('valueChange')
  onInoInputValueChange(ev: CustomEvent<string>): void {
    if ((ev.target as HTMLElement).tagName.toLowerCase() === 'ino-input') {
      ev.stopImmediatePropagation();
      this.inputChanged(ev.detail);
    }
  }

  @Listen('keydown')
  onKeyDown(ev: KeyboardEvent): void {
    if (
      !this.filteredOptionEls ||
      this.filteredOptionEls.length === 0 ||
      (this.lastKeyCode !== 'Enter' && !this.menuIsVisible)
    ) {
      return;
    }

    if (this.lastKeyCode === 'Enter' && !this.menuIsVisible) {
      this.openMenu();
    }
    this.lastKeyCode = ev.code;

    switch (ev.code) {
      case 'Enter':
        this.onEnterPress();
        break;
      case 'ArrowDown':
        this.onArrowDownPress();
        this.scroll(ev);
        break;
      case 'ArrowUp':
        this.onArrowUpPress();
        this.scroll(ev);
        break;
    }
  }

  private onEnterPress = (): void => {
    this.onOptionSelect();
    this.inputEl.getInputElement().then(moveCursorToEnd);
  };

  private onOptionSelect = (): void => {
    if (!this.selectedOption) {
      return;
    }

    this.inputChanged(this.selectedOption.innerText);
    this.emitValueOfSelectedOption();
    this.closeMenu();
  };

  private onArrowDownPress = (): void => {
    const nextIndex =
      this.filteredOptionEls.findIndex(
        (o) => o.value === this.selectedOption?.value
      ) + 1;
    const isIndexOutOfBound = nextIndex >= this.filteredOptionEls.length;
    const filteredOptionIndex = isIndexOutOfBound ? 0 : nextIndex;
    this.selectedOptionIndex = this.optionEls.indexOf(
      this.filteredOptionEls[filteredOptionIndex]
    );
  };

  private onArrowUpPress = (): void => {
    const nextIndex =
      this.filteredOptionEls.findIndex(
        (o) => o.value === this.selectedOption?.value
      ) - 1;
    const isIndexOutOfBound = nextIndex < 0;
    const filteredOptionIndex = isIndexOutOfBound
      ? this.filteredOptionEls.length - 1
      : nextIndex;
    this.selectedOptionIndex = this.optionEls.indexOf(
      this.filteredOptionEls[filteredOptionIndex]
    );
  };

  private scroll = (ev: KeyboardEvent): void => {
    ev.preventDefault();
    const { offsetTop, clientHeight: optionHeight } = this.selectedOption;
    const { scrollTop, clientHeight } = this.menuContainer;
    if (
      offsetTop + optionHeight / 2 > scrollTop + clientHeight ||
      offsetTop < scrollTop
    ) {
      this.menuContainer.scrollTo(0, offsetTop);
    }
  };

  private setupInput = (): void => {
    if (!hasSlotContent(this.el, 'input')) {
      throw new Error(
        'The slot "input" is empty. Please provide an ino-input element to that slot.'
      );
    }

    this.inputEl = getSlotContent(this.el, 'input') as HTMLInoInputElement;

    if (this.inputEl.value) {
      console.warn(
        'The value property of the <ino-input> is managed by the <ino-autocomplete> and therefore should not be set manually.'
      );
    }

    // Workaround because input does not throw native focus/blur events
    this.inputEl.addEventListener('inoFocus', this.onInoInputFocus);
    this.inputEl.addEventListener('inoBlur', this.onInoInputBlur);

    // ensure same width as the input
    this.menuContainer.style.setProperty(
      '--input-width',
      window.getComputedStyle(this.inputEl).width
    );
  };

  private onInoInputFocus = (): void => {
    if (!this.inputEl.disabled) {
      this.openMenu();
    }
  };

  private onInoInputBlur = (event: CustomEvent<FocusEvent>): void => {
    if (this.isOptionClick(event.detail)) {
      return;
    }
    this.closeMenu();

    const newIndex = this.optionEls.findIndex(
      (option) => option.innerText.trim() === this.inputEl.value.trim()
    );

    if (newIndex !== this.selectedOptionIndex) {
      this.selectedOptionIndex = newIndex;
      this.emitValueOfSelectedOption();
    }
    this.inputChanged(this.selectedOption?.innerText || '');
  };

  /**
   * Checks if the newly focused element is a list item.
   * This happen e.g. on a list item click.
   */
  private isOptionClick = (ev: FocusEvent): boolean =>
    ev.relatedTarget &&
    ((ev.relatedTarget as HTMLElement).matches('.mdc-list-item') ||
      (ev.relatedTarget as HTMLElement).matches('ino-option'));

  private emitValueOfSelectedOption = (): void => {
    this.valueChange.emit(this.selectedOption?.value);
  };

  private inputChanged = (searchTerm: string): void => {
    if (!this.inputEl) {
      return;
    }

    this.inputEl.value = searchTerm || '';
    this.debouncer.debounce(
      () => this.filterOptions(this.inputEl.value),
      this.debounceTimeout
    );
  };

  private filterOptions = (newVal: string): void => {
    this.filteredOptionEls = this.optionEls.filter((option) => {
      const matched = option.innerText
        .trim()
        .toLowerCase()
        .includes(newVal.trim().toLowerCase());
      option.style.display = matched ? 'block' : 'none';
      return matched;
    });
    this.noOptionsIsVisible = this.filteredOptionEls.length === 0;
  };
  private setupOptions = (): void => {
    this.optionEls = Array.from(this.el.getElementsByTagName('ino-option'));
    this.filteredOptionEls = this.optionEls;
    this.setOptionByValue(this.value);
  };

  private setOptionByValue = (value: string): void => {
    if (!this.optionEls) {
      return;
    }
    this.selectedOptionIndex = this.optionEls.findIndex(
      (oEl) => oEl.value === value
    );
    this.inputChanged(this.selectedOption?.innerText);
  };

  // necessary because the input's blur will trigger BEFORE click but AFTER mousedown
  private onListItemClick = (ev: MouseEvent): void => {
    const { parentElement } = ev.target as HTMLElement;
    if (
      parentElement.nodeName.toLowerCase() !== 'ino-option' &&
      parentElement.parentElement.nodeName.toLowerCase() !== 'ino-option'
    ) {
      return;
    }

    const inoOption = (parentElement.nodeName.toLowerCase() === 'ino-option'
      ? parentElement
      : parentElement.parentElement) as HTMLInoOptionElement;

    this.selectedOptionIndex = this.optionEls.findIndex(
      (oEl) => oEl.value === inoOption.value
    );
    this.onOptionSelect();
  };

  private openMenu = () => (this.menuIsVisible = true);
  private closeMenu = () => (this.menuIsVisible = false);

  render() {
    const menuClasses = classNames({
      menu: true,
      'menu-hidden': !this.menuIsVisible,
      'menu-shown': this.menuIsVisible,
      'mdc-list': true,
    });

    return (
      <Host>
        <slot name="input" />
        <div
          class={menuClasses}
          ref={(el) => (this.menuContainer = el)}
          onMouseDown={(ev) => this.onListItemClick(ev)}
        >
          {this.noOptionsIsVisible && (
            <p class="no-options-text">{this.noOptionsText}</p>
          )}
          <slot onSlotchange={this.setupOptions} />
        </div>
      </Host>
    );
  }
}
