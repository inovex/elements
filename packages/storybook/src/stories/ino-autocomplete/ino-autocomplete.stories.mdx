import { Meta, Description } from "@storybook/addon-docs";
import readme from "../../../../elements/src/components/ino-autocomplete/readme.md";

<Meta title="Input/ino-autocomplete/Documentation" />

<Description markdown={readme} />
