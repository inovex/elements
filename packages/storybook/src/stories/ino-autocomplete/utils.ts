import { KeyValue } from '@inovex.de/elements';

export const AutoCompleteStoryOptions: KeyValue[] = [
  {
    key: 'hh',
    value: 'Hamburg',
  },
  {
    key: 'ber',
    value: 'Berlin',
  },
  {
    key: 'k',
    value: 'Karlsruhe',
  },
];
