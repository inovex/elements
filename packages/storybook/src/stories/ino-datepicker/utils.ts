import '@inovex.de/elements';

export function registerInlineDatepickerHandler() {
  const group = document.querySelector('ino-radio-group') as HTMLInoRadioGroupElement;
  const datepicker = document.querySelector('.datepicker-group ino-datepicker') as HTMLInoDatepickerElement;

  const today = document.querySelector('.today');
  const thisWeek = document.querySelector('.thisWeek');
  const lastWeek = document.querySelector('.lastWeek');

  const todayClickHandler = () => {
    datepicker.value = '02.03.2020';
    datepicker.label = 'Am';
    group.value = 'at';
    datepicker.range = false;
  };

  const thisWeekClickHandler = () => {
    group.value = 'range';
    datepicker.label = 'Zeitraum';
    datepicker.range = true;
    datepicker.value = '02.03.2020 bis 08.03.2020';
  };

  const lastWeekClickHandler = () => {
    group.value = 'range';
    datepicker.label = 'Zeitraum';
    datepicker.range = true;
    datepicker.value = '24.02.2020 bis 01.03.2020';
  };

  const checkedChangeHandler = (e: CustomEvent<boolean>) => {
    const target = e.target as HTMLInoRadioElement;

    switch (target.value) {
      case 'at':
        datepicker.label = 'Am';
        datepicker.range = false;
        break;
      case 'after':
        datepicker.label = 'Ab';
        datepicker.range = false;
        if (group.value == 'range') {
          datepicker.value = datepicker.value.split(' - ').shift();
        }
        break;
      case 'before':
        datepicker.label = 'Bis';
        datepicker.range = false;
        if (group.value == 'range') {
          datepicker.value = datepicker.value.split(' - ').pop();
        }
        break;
      case 'range':
        datepicker.range = true;
        datepicker.label = 'Zeitraum';
        break;
    }
    group.value = target.value;
  };

  today.addEventListener('click', todayClickHandler);
  thisWeek.addEventListener('click', thisWeekClickHandler);
  lastWeek.addEventListener('click', lastWeekClickHandler);
  group.addEventListener('checkedChange', checkedChangeHandler);

  return () => {
    today.removeEventListener('click', todayClickHandler);
    thisWeek.removeEventListener('click', thisWeekClickHandler);
    lastWeek.removeEventListener('click', lastWeekClickHandler);
    group.removeEventListener('checkedChange', checkedChangeHandler);
  };
}
