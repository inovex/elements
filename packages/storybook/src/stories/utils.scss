/**
 * Applies styling to the container of a story.
 * The styling is encapsulated within this mixin because the ID may change in the future.
 */
@mixin story-container {
  #root-inner {
    @content;
  }
}
