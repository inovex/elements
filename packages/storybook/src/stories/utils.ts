import { html } from 'lit-html';
import ICONS from '../../../elements/src/components/ino-icon/icons';

export const maybeCreateStoryArgs = (story: any) => {
  if (!story.args) {
    story.args = {};
  }
  if (!story.argTypes) {
    story.argTypes = {};
  }
};

export const withIconControl = (
  story: any,
  propertyName: string,
  defaultValue?: string
) => {
  maybeCreateStoryArgs(story);
  story.args[propertyName] = defaultValue || '';
  story.argTypes[propertyName] = {
    control: {
      type: 'select',
    },
    options: getIcons(),
  };
};

export const withColorScheme = (
  story: any,
  propertyName: string,
  defaultValue?: string
) => {
  maybeCreateStoryArgs(story);
  story.args[propertyName] = defaultValue || '';
  story.argTypes[propertyName] = {
    control: {
      type: 'select',
    },
    options: [
      'primary',
      'secondary',
      'success',
      'warning',
      'error',
      'light',
      'dark',
    ],
  };
};

export const withSortDirection = (
  story: any,
  propertyName: string,
  defaultValue?: string
) => {
  maybeCreateStoryArgs(story);
  story.args[propertyName] = defaultValue || '';
  story.argTypes[propertyName] = {
    options: ['desc', 'asc'],
    control: {
      type: 'select',
    },
  };
};

export const getIcons = () => ICONS.filter((icon) => !icon.startsWith('_'));

export const defaultDecorator = (story, className?: string) => {
  return html` <div class="${className ?? ''}">${story()}</div> `;
};
