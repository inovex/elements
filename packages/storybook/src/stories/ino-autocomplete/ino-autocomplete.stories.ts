import { Components } from '@inovex.de/elements';
import { useEffect } from '@storybook/client-api';
import { Story } from '@storybook/web-components';
import { html } from 'lit-html';
import { defaultDecorator, showSnackbar } from '../utils';

import './ino-autocomplete.scss';

export default {
  title: 'Input/ino-autocomplete',
  component: 'ino-autocomplete',
  parameters: {
    actions: {
      handles: ['optionSelected'],
    },
  },
  decorators: [
    defaultDecorator,
    (story) => {
      useEffect(() => {
        const eventHandler = function (e: any) {
          showSnackbar(`"${e.detail}" was selected.`);
        };

        document.addEventListener('optionSelected', eventHandler);

        return () => {
          document.removeEventListener('optionSelected', eventHandler);
        };
      });

      return story();
    },
  ],
};

export const Playground: Story<Components.InoAutocomplete> = (args) => html`
  <div style="height: 300px;">
    <ino-autocomplete
      style="margin: 50px"
      debounce-timeout="${args.debounceTimeout}"
    >
      <ino-input slot="input" />
      <ino-list slot="list">
        <ino-list-item text="Item A"></ino-list-item>
        <ino-list-item text="Item B"></ino-list-item>
        <ino-list-item text="Item C"></ino-list-item>
      </ino-list>
    </ino-autocomplete>
  </div>
`;

Playground.args = {
  debounceTimeout: 300,
};
