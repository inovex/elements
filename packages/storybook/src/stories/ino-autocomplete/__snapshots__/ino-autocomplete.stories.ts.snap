// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input/ino-autocomplete Playground smoke-test 1`] = `
<div id="root-inner">
  <div class>
    <div style="height: 300px;">
      <ino-autocomplete style="margin: 50px"
                        no-options-text="No option found"
                        class="hydrated"
      >
        <ino-input slot="input"
                   helper="Helper text"
                   helper-persistent
                   class="hydrated"
                   tabindex="-1"
        >
          <label class="ino-input__composer mdc-text-field mdc-text-field--filled mdc-text-field--box mdc-text-field--no-label mdc-ripple-upgraded">
            <ino-label class="hydrated">
              <div class="mdc-line-ripple">
              </div>
            </ino-label>
            <input class="mdc-text-field__input"
                   step="1"
                   type="text"
                   aria-controls="input-helper-text__0"
                   aria-describedby="input-helper-text__0"
            >
          </label>
          <div class="mdc-text-field-helper-line">
            <div class="mdc-text-field-helper-text mdc-text-field-helper-text--persistent"
                 id="input-helper-text__0"
                 aria-hidden="true"
            >
              Helper text
            </div>
          </div>
        </ino-input>
        <ino-option value="value of Option 0"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 0"
          >
            <span class="mdc-deprecated-list-item__text">
              0
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 1"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 1"
          >
            <span class="mdc-deprecated-list-item__text">
              1
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 2"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 2"
          >
            <span class="mdc-deprecated-list-item__text">
              2
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 3"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 3"
          >
            <span class="mdc-deprecated-list-item__text">
              3
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 4"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 4"
          >
            <span class="mdc-deprecated-list-item__text">
              4
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 5"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 5"
          >
            <span class="mdc-deprecated-list-item__text">
              5
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 6"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 6"
          >
            <span class="mdc-deprecated-list-item__text">
              6
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 7"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 7"
          >
            <span class="mdc-deprecated-list-item__text">
              7
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 8"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 8"
          >
            <span class="mdc-deprecated-list-item__text">
              8
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 9"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 9"
          >
            <span class="mdc-deprecated-list-item__text">
              9
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 10"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 10"
          >
            <span class="mdc-deprecated-list-item__text">
              10
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 11"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 11"
          >
            <span class="mdc-deprecated-list-item__text">
              11
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 12"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 12"
          >
            <span class="mdc-deprecated-list-item__text">
              12
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 13"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 13"
          >
            <span class="mdc-deprecated-list-item__text">
              13
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 14"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 14"
          >
            <span class="mdc-deprecated-list-item__text">
              14
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 15"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 15"
          >
            <span class="mdc-deprecated-list-item__text">
              15
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 16"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 16"
          >
            <span class="mdc-deprecated-list-item__text">
              16
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 17"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 17"
          >
            <span class="mdc-deprecated-list-item__text">
              17
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 18"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 18"
          >
            <span class="mdc-deprecated-list-item__text">
              18
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 19"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 19"
          >
            <span class="mdc-deprecated-list-item__text">
              19
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 20"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 20"
          >
            <span class="mdc-deprecated-list-item__text">
              20
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 21"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 21"
          >
            <span class="mdc-deprecated-list-item__text">
              21
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 22"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 22"
          >
            <span class="mdc-deprecated-list-item__text">
              22
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 23"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 23"
          >
            <span class="mdc-deprecated-list-item__text">
              23
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 24"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 24"
          >
            <span class="mdc-deprecated-list-item__text">
              24
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 25"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 25"
          >
            <span class="mdc-deprecated-list-item__text">
              25
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 26"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 26"
          >
            <span class="mdc-deprecated-list-item__text">
              26
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 27"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 27"
          >
            <span class="mdc-deprecated-list-item__text">
              27
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 28"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 28"
          >
            <span class="mdc-deprecated-list-item__text">
              28
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 29"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 29"
          >
            <span class="mdc-deprecated-list-item__text">
              29
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 30"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 30"
          >
            <span class="mdc-deprecated-list-item__text">
              30
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 31"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 31"
          >
            <span class="mdc-deprecated-list-item__text">
              31
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 32"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 32"
          >
            <span class="mdc-deprecated-list-item__text">
              32
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 33"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 33"
          >
            <span class="mdc-deprecated-list-item__text">
              33
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 34"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 34"
          >
            <span class="mdc-deprecated-list-item__text">
              34
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 35"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 35"
          >
            <span class="mdc-deprecated-list-item__text">
              35
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 36"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 36"
          >
            <span class="mdc-deprecated-list-item__text">
              36
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 37"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 37"
          >
            <span class="mdc-deprecated-list-item__text">
              37
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 38"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 38"
          >
            <span class="mdc-deprecated-list-item__text">
              38
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 39"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 39"
          >
            <span class="mdc-deprecated-list-item__text">
              39
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 40"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 40"
          >
            <span class="mdc-deprecated-list-item__text">
              40
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 41"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 41"
          >
            <span class="mdc-deprecated-list-item__text">
              41
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 42"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 42"
          >
            <span class="mdc-deprecated-list-item__text">
              42
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 43"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 43"
          >
            <span class="mdc-deprecated-list-item__text">
              43
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 44"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 44"
          >
            <span class="mdc-deprecated-list-item__text">
              44
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 45"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 45"
          >
            <span class="mdc-deprecated-list-item__text">
              45
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 46"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 46"
          >
            <span class="mdc-deprecated-list-item__text">
              46
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 47"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 47"
          >
            <span class="mdc-deprecated-list-item__text">
              47
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 48"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 48"
          >
            <span class="mdc-deprecated-list-item__text">
              48
            </span>
          </li>
        </ino-option>
        <ino-option value="value of Option 49"
                    class="hydrated"
                    style="display: block;"
        >
          <li class="mdc-deprecated-list-item"
              data-value="value of Option 49"
          >
            <span class="mdc-deprecated-list-item__text">
              49
            </span>
          </li>
        </ino-option>
      </ino-autocomplete>
    </div>
  </div>
</div>
`;
